client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Selects nodes that can receive requests. Use with
+ * {@link RestClientActions#withNodeSelector withNodeSelector}.
+ */
+public interface NodeSelector {
+    /**
+     * Select the {@link Node}s to which to send requests.
+     *
+     * @param nodes an unmodifiable list of {@linkplain Node}s in the order
+     *      that the {@link RestClient} would prefer to use them
+     * @return a subset of the provided list of {@linkplain Node}s that the
+     *      selector approves of, in the order that the selector would prefer
+     *      to use them.
+     */
+    List<Node> select(List<Node> nodes);
+
+    /**
+     * Selector that matches any node.
+     */
+    NodeSelector ANY = new NodeSelector() {
+        @Override
+        public List<Node> select(List<Node> nodes) {
+            return nodes;
+        }
+
+        @Override
+        public String toString() {
+            return "ANY";
+        }
+    };
+
+    /**
+     * Selector that matches any node that has metadata and doesn't
+     * have the {@code master} role OR it has the data {@code data}
+     * role. It does not reorder the nodes sent to it.
+     */
+    NodeSelector NOT_MASTER_ONLY = new NodeSelector() {
+        @Override
+        public List<Node> select(List<Node> nodes) {
+            List<Node> subset = new ArrayList<>(nodes.size());
+            for (Node node : nodes) {
+                if (node.getRoles() == null) continue;
+                if (false == node.getRoles().hasMasterEligible() || node.getRoles().hasData()) {
+                    subset.add(node);
+                }
+            }
+            return subset;
+        }
+
+        @Override
+        public String toString() {
+            return "NOT_MASTER_ONLY";
+        }
+    };
+
+    /**
+     * Selector that composes two selectors, running the "right" most selector
+     * first and then running the "left" selector on the results of the "right"
+     * selector.
+     */
+    class Compose implements NodeSelector {
```

Comment:
I am pretty sure that we don't need this, especially given that the only time we use it in our tests is to compose (any , selector) . It is cool but I think it adds complexity, needs testing, while a single selector can do all it wants alone, no need to chain them.

Reply:
We also use it in `RestClientView.withNodeSelector`. I expect we want to keep `withNodeSelector` on `RestClientView` because we'll end up with other `withXXX` style methods which will all return `RestClientView`. So folks will do `client.withTime(123).withNodeSelector(bar)`.

We could implement `RestClientView.withNodeSelector` to check if the `NodeSelector` in the view is `ANY` and throw an `IllegalStateException` if it isn't but that feels weird.

Reply:
I would keep things simple and use `SetOnce` to make sure that users can only set one node selector, if not set we use `ANY`

Reply:
oh man SetOnce is a lucene thing though :x